---
title: "dynamic_allelic_changes"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{dynamic_allelic_changes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Allelic dynamic changes

### Introduction
To identify genes that change their allelic usage over developmental time, ASPEN requires cell to be assigned a descrete time value. If sequencing experiments were done across a number of time points, this is simply identities of the respective experiment batch. Please note, that the batch correction might be required to remove any technical batch effects. Alternatively, cells can be ordered by the developmental trajectory through pseudotime analysis (with tools like Monocle, Slingshot, palantir, etc.). Pseudotime is a probability of a cell to be in a terminal state, which is restricted to $[0,1]$ interval. This vector can be divided into equal-sized bins and the cells are grouped together. \\
To test for changes in the mean allelic ratio between the discrete time points, ASPEN evaluates whether a gene's allelic distribution is the same across the time points ($H_0$) or the allelic distribution is different ($H_1$).  ASPEN goes through the following steps:\\
\begin{enumerate}
\item 1) estimate beta-binomial distribution parameters. For each gene, beta-binomial $\alpha$ and $\beta$ parameters are estimated by maximum likelihood estimation (MLE). Using $\alpha$ and $\beta$, we can calculate allelic ratio mean, $\mu$, and dispersion, $\theta$ as
\\ $\mu=\frac{\alpha}{\alpha + \beta}$, and
\\ $\theta = \frac{1}{\alpha + \beta}$;\\
This is done for each discrete group separately and using all cell (without subdividing them into groups);
\item 2) fit a local regression for $\theta$ as a function of total gene expression - again this step is performed for each discrete group and for all cells together;
\item 3) estimate shrinkage parameters, $\delta$ and $N$, appropriate for your data - same as above;
\item 4) shrink the original $\theta$ towards the common trend - same as above;
\item 5) perform test for dynamic allelic imbalance.
\end{enumerate}

### Setup
```{r setup}
#loading required libraries
library(ASPEN)
library(gridExtra)
library(openxlsx)
```

### Loading allele-specifc count data
Like for other vignettes, we use mouse brain organoids data from CastB6 hybrids. The pseudotime for this data was estimated using palantir and we select three cell types, representing an early neurodevelopment: radial glial cells (RGCs), intermediate progenitors cells (IPCs), deep layer neurons (cortical neurons). 


```{r}
data("Cast_B6_a1")
data("Cast_B6_tot")
load_file <- system.file("extdata", "Cast_B6_cell_annot.xlsx", package = "ASPEN")
cell_annot <- read.xlsx(load_file, rowNames = T)
#loading pseudotime assignment
load_time <- system.file("extdata", "pseudotime_CastB6.xlsx", package = "ASPEN")
pseudotime <- read.xlsx(load_time, rowNames = T)
head(pseudotime)
```

We select the barcodes for the cell types that have pseudotime estimates. 
```{r}

Cast_B6_a1 <- Cast_B6_a1[,colnames(Cast_B6_a1) %in% pseudotime$cell_id]
Cast_B6_tot <- Cast_B6_tot[,colnames(Cast_B6_tot) %in% pseudotime$cell_id]

```


Remove low-expressed genes that have counts in less than 10 cells
```{r}
#Remove low-expressed genes which have less than 10 cells 
Cast_B6_tot  <- Cast_B6_tot[rowSums(Cast_B6_tot > 1) >= 10,]
Cast_B6_a1  <- Cast_B6_a1[rownames(Cast_B6_tot),]

```



Dividing pseudotime vector into five equal-sized groups
```{r}
pseudotime <- pseudotime[match(colnames(Cast_B6_tot), pseudotime$cell_id),]
pseudotime$group <- cut(pseudotime$time, breaks=c(quantile(pseudotime$time, probs = seq(0, 1, by = 0.2))))
pseudotime$group[is.na(pseudotime$group)] <- levels(pseudotime$group)[1]
#adding cell ids to pseudotime obejct row names
rownames(pseudotime) <- pseudotime$cell_id
head(pseudotime)

```

```{r}
table(pseudotime$group)
```

```{r}
dim(Cast_B6_a1)
dim(Cast_B6_tot)
table(rownames(Cast_B6_a1) == rownames(Cast_B6_tot))
```


### Estimating beta-binomial parameters
We start by estimating beta-binomial distribution parameters for each gene across all cells. Theu will be used for the likelihood calculation under $H_0$ - assumes no differences over pseudotime.
```{r}
global_params <- estim_bbparams(Cast_B6_a1, Cast_B6_tot, min_cells = 5, cores = 6)
global_shrunk <- correct_theta_bygroup(global_params, delta_set = 50, N_set = 30, thetaFilter = 0.001)
```

Checking that the order of genes in the count matrix and the parameter estimates object is the same.
```{r}
table(rownames(Cast_B6_a1) == rownames(global_shrunk))
```


Visualizing the local model fit and the shrunk dispersion estimates
```{r, fig.width=6.5, fig.height=3}
fit_plot <- plot_disp_fit_theta(global_shrunk, midpoint = 200)
shrunk_plot <- plot_disp(global_shrunk) + 
  geom_hline(yintercept = log(1e-03), linetype = "dashed", linewidth = 1)
  
grid.arrange(fit_plot, shrunk_plot, ncol = 2)
```

Under alternative hypothesis, we assume that the alellic distributions are different between the time points, so we repeat the estimation and dispersion shrinkage steps for each group separately 

```{r}
#Estimating distribution parameters
group_params <- estim_bbparams_bygroup(Cast_B6_a1, Cast_B6_tot, metadata = pseudotime, split.var = "group", min_cells = 5, cores = 6)

#removing groups where optim did not converge
group_params <- group_params[!is.na(group_params$bb_theta),]
group_params <- as.data.frame(group_params)
shrunk_group_params <- correct_theta_bygroup(group_params, delta_set = 50, N_set = 30, thetaFilter = 0.001)
```


Visualizing the local model fit when each group-level observations are treated as independent variables.

```{r, fig.width=6.5, fig.height=3}
fit_plot_group <- plot_disp_fit_theta(shrunk_group_params, midpoint = 800) +
    ylim(-10, 4)
shrunk_plot_group <- plot_disp(shrunk_group_params) + 
  geom_hline(yintercept = log(1e-03), linetype = "dashed", linewidth = 1) +
   ylim(-10, 4)
  
grid.arrange(fit_plot_group, shrunk_plot_group, ncol = 2)
```

### Test for changes in allelic distribution over time

ASPEN requires the estimation for each gene to be provided as a separate data frame. This can be done by splitting the group estimates object by gene.
```{r}
group_params_list <- split(shrunk_group_params, f = shrunk_group_params$gene)
#re-ordering to match order of genes in the count matrices 
group_params_list <- group_params_list[rownames(Cast_B6_tot)]

#Checking that the order of genes in the count matrix and the group-level parameter estimates object is the same
table(names(group_params_list) == rownames(Cast_B6_tot))
```

Checking that the order of genes in global and group-level  parameter estimates objects are the same.
```{r}
table(rownames(global_shrunk) == names(group_params_list))
```


```{r}
pseudot_change <- allelicSwitch(Cast_B6_a1, Cast_B6_tot, min_cells = 5, metadata = pseudotime,  split.var = "group", estimates = global_shrunk, estimates_group = group_params_list)

```

For genes in which the quality cut-off threshold (here we used a minimum of five cells with at least of five mapped reads) was not met, the inference is not performed, Those genes have NA values in the fields specified above. We remove those genes and calculate fdr's. 
```{r}
pseudot_change <- pseudot_change[!is.na(pseudot_change$pval),]

#at which time point the variation in allelic ratio's deviates from the expected levels
#p-values for each time bin are the last five column, so p.adjust is applied to all of them 
pseudot_change[,35:39] <- lapply(pseudot_change[,30:34], function(q) p.adjust(q, method = "fdr")) 
colnames(pseudot_change)[35:39] <- paste0("fdr_bin", seq(1:5))
#allelic distribution changes over time
pseudot_change$fdr_mean <- p.adjust(pseudot_change$pval, method = "fdr")
#variance changes overtime, whilst keeping the mean AR constant
pseudot_change$fdr_var <- p.adjust(pseudot_change$pval_var, method = "fdr")


```


Top 10 genes with significant changes in allelic distribution over pseudotime
```{r}
pseudot_change <- pseudot_change[order(pseudot_change$fdr_mean),]
pseudot_change[1:10, c("mu_global", colnames(pseudot_change)[grep("mu_group", colnames(pseudot_change))], "fdr_mean", "fdr_var")]
```

We can pick the genes that only change in their mean allelic ratio over pseudotime. For such genes we set fdr_var > 0.05.
```{r}
head(pseudot_change[pseudot_change$fdr_var > 0.05, c("mu_global", colnames(pseudot_change)[grep("mu_group", colnames(pseudot_change))], "fdr_mean", "fdr_var")], n = 10)
```


Visualizing changes along pseudotime

```{r, fig.width=4, fig.height=4.5}
gene <- "Cst3"
#generating data frame for plotting
#plot_data <- makedf(Cast_B6_a1, Cast_B6_tot, gene = gene, metadata = pseudotime, order.by = "time", split.var = "group")

simul_data <- make_plotdf_simul(Cast_B6_a1, Cast_B6_tot, gene = gene, estimates_group = group_params_list,
                                metadata = pseudotime, order.by = "time", split.var = "group")

plot_distr(simul_data, gene = gene, add.density = FALSE, min_counts = 0) + 
  geom_hline(yintercept = c(simul_data$Index[match(unique(simul_data$group), simul_data$group)][-1])) +
  labs(y = "Ordered pseudotime")
```

Boxplot to assess ASE at each time point. Differences in the mean values are confirmed with Wilcox two-sided test. 
```{r, fig.width=4, fig.height=4.5}
geBoxplot(simul_data, allele1 = "Bl6", allele2 = "Cast")
```

Plotting allelic ratio density using original data and the original estimated theta and common theta and allelic ratio distribution from data simulated from the common theta.
```{r,  fig.width=4, fig.height=4.5}
plot_theta_density(simul_data)
```

There is strong support that the observed ASE dispersion deviates from the expected dispersion.
```{r}
pseudot_change[gene, colnames(pseudot_change)[grep("fdr_bin", colnames(pseudot_change))]]
```

Visualizing the distance from the common dispersion trend for all pseudotime group.
```{r, fig.width=4, fig.height=4.5}
plot_exp_disp(as.data.frame(shrunk_group_params), gene = gene) +
  ylim(-10, 4)
```

Alternatively, we can select genes that have significant changes in variance but not in the mean allelic ratio. 
```{r}
pseudot_change <- pseudot_change[order(pseudot_change$fdr_var),]
head(pseudot_change[pseudot_change$fdr_mean > 0.05, c("mu_global", colnames(pseudot_change)[grep("mu_group", colnames(pseudot_change))], "fdr_mean", "fdr_var")], n = 10)
```

```{r, fig.width=4, fig.height=4.5}
gene <- "Eno1"
#generating data frame for plotting
#plot_data <- makedf(Cast_B6_a1, Cast_B6_tot, gene = gene, metadata = pseudotime, order.by = "time", split.var = "group")

simul_data <- make_plotdf_simul(Cast_B6_a1, Cast_B6_tot, gene = gene, estimates_group = group_params_list,
                                metadata = pseudotime, order.by = "time", split.var = "group")

plot_distr(simul_data, gene = gene, add.density = FALSE, min_counts = 0) + 
  geom_hline(yintercept = c(simul_data$Index[match(unique(simul_data$group), simul_data$group)][-1])) +
  labs(y = "Ordered pseudotime")
```

Boxplot to assess ASE at each time point. Differences in the mean values are confirmed with Wilcox two-sided test. 
```{r, fig.width=4, fig.height=4.5}
geBoxplot(simul_data, allele1 = "Bl6", allele2 = "Cast")
```

Plotting allelic ratio density using original data and the original estimated theta and common theta and allelic ratio distribution from data simulated from the common theta.
```{r,  fig.width=4, fig.height=4.5}
plot_theta_density(simul_data)
```

```{r}
pseudot_change[gene, colnames(pseudot_change)[grep("fdr_bin", colnames(pseudot_change))]]
```


```{r, fig.width=4, fig.height=4.5}
plot_exp_disp(as.data.frame(shrunk_group_params), gene = gene) +
  ylim(-10, 4)
```

Based on the test results that tells us at which time point the observed dispersion deviates from the expected, we can select genes where the change occurs at the later time points.

```{r}
pseudot_change <- pseudot_change[order(pseudot_change$fdr_bin5),]
head(pseudot_change[, c("mu_global", colnames(pseudot_change)[grep("mu_group", colnames(pseudot_change))], "fdr_mean", "fdr_var", colnames(pseudot_change)[grep("fdr_bin", colnames(pseudot_change))])], n = 20)
```

We examine gene Stmn4 where the significant changes in dispersion occurs at time bin 4 and 5.
```{r}
gene <- "Stmn4"
pseudot_change[gene, colnames(pseudot_change)[grep("fdr_bin", colnames(pseudot_change))]]
```



```{r, fig.width=4, fig.height=4.5}

simul_data <- make_plotdf_simul(Cast_B6_a1, Cast_B6_tot, gene = gene, estimates_group = group_params_list,
                                metadata = pseudotime, order.by = "time", split.var = "group")

plot_distr(simul_data, gene = gene, add.density = FALSE, min_counts = 0) + 
  geom_hline(yintercept = c(simul_data$Index[match(unique(simul_data$group), simul_data$group)][-1])) +
  labs(y = "Ordered pseudotime")
```

Boxplot to assess ASE at each time point. Differences in the mean values are confirmed with Wilcox two-sided test. 
```{r, fig.width=4, fig.height=4.5}
geBoxplot(simul_data, allele1 = "Bl6", allele2 = "Cast")
```

```{r,  fig.width=4, fig.height=4.5}
plot_theta_density(simul_data)
```



```{r, fig.width=4, fig.height=4.5}
plot_exp_disp(as.data.frame(shrunk_group_params), gene = gene) +
  ylim(-10, 4)
```
