<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>allelic_imbalance • ASPEN</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="allelic_imbalance">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ASPEN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/allelic_imbalance.html">allelic_imbalance</a>
    </li>
    <li>
      <a href="../articles/allelic_variance.html">allelic_variance</a>
    </li>
    <li>
      <a href="../articles/dynamic_allelic_changes.html">dynamic_allelic_changes</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>allelic_imbalance</h1>
            
      

      <div class="hidden name"><code>allelic_imbalance.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="allelic-imbalance-test">Allelic imbalance test<a class="anchor" aria-label="anchor" href="#allelic-imbalance-test"></a>
</h2>
<div class="section level4">
<h4 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h4>
<p>ASPEN represents a comprehensive statistical framework that is
designed to evaluate patterns in allele-specific expression (ASE). ASPEN
uses beta-binomial distribution to model allelic ratio, a fraction of
reads from a reference allele to the total number of reads mapped to a
gene. Allelic ratio is restricted to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[0,1]</annotation></semantics></math>
interval and their distribution can be described with beta-binomial
mean,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>μ</mi><annotation encoding="application/x-tex">\mu</annotation></semantics></math>,
and dispersion,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>θ</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>.\</p>
<p>scRNA counts, like any sequencing counts, suffer from the
mean-variance bias - we observe greater variability in low-expressed
genes than in high-expressed genes. ASPEN mitigates that bias by
modeling allelic dispersion,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>θ</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>,
as a function of gene expression and uses hierarchial Bayes model to
shrink the original dispersion estimates towards the common trend, which
represents the level of dispersion expected for genes with similar
expression.\</p>
ASPEN input are the reference allele counts and the total counts (a sum
of counts from both alleles) matrices. The main steps in the workflow as
follows:
</div>
<div class="section level4">
<h4 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#loading required libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ASPEN</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: doParallel</span></span>
<span><span class="co">#&gt; Loading required package: foreach</span></span>
<span><span class="co">#&gt; Loading required package: iterators</span></span>
<span><span class="co">#&gt; Loading required package: parallel</span></span>
<span><span class="co">#&gt; Loading required package: locfit</span></span>
<span><span class="co">#&gt; locfit 1.5-9.9    2024-03-01</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="co">#&gt; Loading required package: ggthemes</span></span>
<span><span class="co">#&gt; Loading required package: ggsci</span></span>
<span><span class="co">#&gt; Loading required package: ggpointdensity</span></span>
<span><span class="co">#&gt; Loading required package: ggExtra</span></span>
<span><span class="co">#&gt; Loading required package: ggrepel</span></span>
<span><span class="co">#&gt; Loading required package: ggpubr</span></span>
<span><span class="co">#&gt; Loading required package: assertthat</span></span>
<span><span class="co">#&gt; Loading required package: zoo</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'zoo'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     as.Date, as.Date.numeric</span></span>
<span><span class="co">#&gt; Loading required package: dplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; Loading required package: data.table</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'data.table'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     between, first, last</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:zoo':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     yearmon, yearqtr</span></span>
<span><span class="co">#&gt; Loading required package: VGAM</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: splines</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'gridExtra'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ycphs.github.io/openxlsx/index.html" class="external-link">openxlsx</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="loading-allele-specifc-count-data">Loading allele-specifc count data<a class="anchor" aria-label="anchor" href="#loading-allele-specifc-count-data"></a>
</h4>
<p>Here we use mouse brain organoids data from CastB6 hybrids. We load
the reference allele counts (B6 counts) and the total counts (a sum of
counts from both alleles). The cells were previously annotated based on
the marker genes expression. There are five cell types: neurogenic
progenitor cells (also, radial glial cells (RGCs)), intermediate
progenitors cells (IPCs), deep layer neurons (cortical neurons),
gliogenic progenitor cells (gliogenic RGCs) and olygodendrocyte
precursor cells (OPCs).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"Cast_B6_a1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"Cast_B6_tot"</span><span class="op">)</span></span>
<span><span class="va">load_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Cast_B6_cell_annot.xlsx"</span>, package <span class="op">=</span> <span class="st">"ASPEN"</span><span class="op">)</span></span>
<span><span class="va">cell_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">read.xlsx</a></span><span class="op">(</span><span class="va">load_file</span>, rowNames <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cell_annot</span><span class="op">)</span></span>
<span><span class="co">#&gt;        cell_barcode background                         cell_type clone_number</span></span>
<span><span class="co">#&gt; 4  AAACCCACAGCAGATG    cast_b6        Gliogenic progenitor cells    3rd_clone</span></span>
<span><span class="co">#&gt; 8  AAACCCATCAGAGCGA    cast_b6        Gliogenic progenitor cells    3rd_clone</span></span>
<span><span class="co">#&gt; 16 AAACGAAGTAGTTAGA    cast_b6       Neurogenic progenitor cells    3rd_clone</span></span>
<span><span class="co">#&gt; 19 AAACGAAGTGTTATCG    cast_b6 Intermediate neuronal progenitors    3rd_clone</span></span>
<span><span class="co">#&gt; 22 AAACGAATCTCGTCAC    cast_b6                Deep layer neurons    3rd_clone</span></span>
<span><span class="co">#&gt; 41 AAAGGATGTATGCTTG    cast_b6 Intermediate neuronal progenitors    3rd_clone</span></span>
<span><span class="co">#&gt;         cell_idents</span></span>
<span><span class="co">#&gt; 4    Gliogenic RGCs</span></span>
<span><span class="co">#&gt; 8    Gliogenic RGCs</span></span>
<span><span class="co">#&gt; 16             RGCs</span></span>
<span><span class="co">#&gt; 19             IPCs</span></span>
<span><span class="co">#&gt; 22 Cortical neurons</span></span>
<span><span class="co">#&gt; 41             IPCs</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cell_annot</span><span class="op">$</span><span class="va">cell_idents</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Cortical neurons   Gliogenic RGCs             IPCs             OPCs </span></span>
<span><span class="co">#&gt;             1615             1157             1274             1523 </span></span>
<span><span class="co">#&gt;             RGCs </span></span>
<span><span class="co">#&gt;              935</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="data-pre-processing">Data pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"></a>
</h4>
<p>Analysis of allelic imblanace will be done in each cell type
separately. First, we split the metadata object by cell types.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#splitting the metadata by cell type</span></span>
<span><span class="va">cell_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">cell_annot</span>, f <span class="op">=</span> <span class="va">cell_annot</span><span class="op">$</span><span class="va">cell_idents</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a1_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cell_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="va">Cast_B6_a1</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*_"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Cast_B6_a1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">q</span><span class="op">$</span><span class="va">cell_barcode</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">tot_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cell_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="va">Cast_B6_tot</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*_"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Cast_B6_tot</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">q</span><span class="op">$</span><span class="va">cell_barcode</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Checking that reference and total count matrices are the same
size.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span>, <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`Cortical neurons`</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Gliogenic RGCs`</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $IPCs</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OPCs</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $RGCs</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span></code></pre></div>
<p>Checking that the gene order is the same between the reference and
total count matrices.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span><span class="op">)</span>, <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`Cortical neurons`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1551 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Gliogenic RGCs`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1551 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $IPCs</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1551 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OPCs</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1551 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $RGCs</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1551</span></span></code></pre></div>
<p>Remove low-expressed genes that have counts in less than 10 cells</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Remove low-expressed genes which have less than 10 cells </span></span>
<span><span class="va">tot_mat</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">tot_mat</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="va">q</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">q</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">a1_mat</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="va">p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span>,<span class="op">]</span>, </span>
<span>                  <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="estimating-beta-binomial-parameters">Estimating beta-binomial parameters<a class="anchor" aria-label="anchor" href="#estimating-beta-binomial-parameters"></a>
</h4>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_init_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/estim_bbparams.html">estim_bbparams</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, min_cells <span class="op">=</span> <span class="fl">5</span>, cores <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>, <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
The output is a table with the following columns:
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bb_init_params</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;           N          AR tot_gene_mean tot_gene_variance        alpha</span></span>
<span><span class="co">#&gt; Stmn1   850 0.984148582     5.9328689         29.297138 1.752689e+02</span></span>
<span><span class="co">#&gt; Uchl1  1055 0.998929294     8.2020266         45.383114 1.735938e+05</span></span>
<span><span class="co">#&gt; Hmgb1    37 0.976436130     1.0880304          1.781219 1.458176e+01</span></span>
<span><span class="co">#&gt; Gm9794   25 0.011239671     0.9550348          1.544872 8.007730e-01</span></span>
<span><span class="co">#&gt; Ftl1   1432 0.694541696    21.2995567        322.520475 1.867524e+02</span></span>
<span><span class="co">#&gt; Mif      94 0.003891258     1.4046865          3.597215 8.830451e+00</span></span>
<span><span class="co">#&gt;                beta  bb_mu bb_theta id</span></span>
<span><span class="co">#&gt; Stmn1     2.4149423 0.9864   0.0056  1</span></span>
<span><span class="co">#&gt; Uchl1   134.0358378 0.9992   0.0000  2</span></span>
<span><span class="co">#&gt; Hmgb1     0.3430136 0.9770   0.0670  3</span></span>
<span><span class="co">#&gt; Gm9794   63.9134233 0.0124   0.0155  4</span></span>
<span><span class="co">#&gt; Ftl1     85.9937477 0.6847   0.0037  5</span></span>
<span><span class="co">#&gt; Mif    2439.8454192 0.0036   0.0004  6</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="estimate-appropriate-shrinkage-parameters">Estimate appropriate shrinkage parameters<a class="anchor" aria-label="anchor" href="#estimate-appropriate-shrinkage-parameters"></a>
</h4>
<p>We estimate shrinkage parameters on each cell type separately</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1001011</span><span class="op">)</span></span>
<span><span class="va">shrink_pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_init_params</span>, <span class="va">estim_delta</span><span class="op">)</span></span>
<span><span class="va">shrink_pars</span></span>
<span><span class="co">#&gt; $`Cortical neurons`</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    13    19 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Gliogenic RGCs`</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    11    17 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $IPCs</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    13    19 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OPCs</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    12    19 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $RGCs</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    15    21</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="performing-bayesian-shrinkage">Performing Bayesian shrinkage<a class="anchor" aria-label="anchor" href="#performing-bayesian-shrinkage"></a>
</h4>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_init_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_init_params</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="va">q</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">q</span><span class="op">$</span><span class="va">bb_theta</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">shrunk_estims_vardelta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/correct_theta.html">correct_theta</a></span><span class="op">(</span><span class="va">p</span>, N_set <span class="op">=</span> <span class="va">q</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, delta_set <span class="op">=</span> <span class="va">q</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, thetaFilter <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span>,</span>
<span>                                 <span class="va">bb_init_params</span>, <span class="va">shrink_pars</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">shrunk_estims_vardelta</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;           N          AR tot_gene_mean tot_gene_variance        alpha</span></span>
<span><span class="co">#&gt; Stmn1   850 0.984148582     5.9328689         29.297138 1.752689e+02</span></span>
<span><span class="co">#&gt; Uchl1  1055 0.998929294     8.2020266         45.383114 1.735938e+05</span></span>
<span><span class="co">#&gt; Hmgb1    37 0.976436130     1.0880304          1.781219 1.458176e+01</span></span>
<span><span class="co">#&gt; Gm9794   25 0.011239671     0.9550348          1.544872 8.007730e-01</span></span>
<span><span class="co">#&gt; Ftl1   1432 0.694541696    21.2995567        322.520475 1.867524e+02</span></span>
<span><span class="co">#&gt; Mif      94 0.003891258     1.4046865          3.597215 8.830451e+00</span></span>
<span><span class="co">#&gt;                beta  bb_mu bb_theta id theta_smoothed  ci_upper    ci_lower</span></span>
<span><span class="co">#&gt; Stmn1     2.4149423 0.9864   0.0056  1    0.007485426 0.1367155 -0.12174465</span></span>
<span><span class="co">#&gt; Uchl1   134.0358378 0.9992   0.0000  2             NA        NA          NA</span></span>
<span><span class="co">#&gt; Hmgb1     0.3430136 0.9770   0.0670  3    0.025846831 0.1106315 -0.05893784</span></span>
<span><span class="co">#&gt; Gm9794   63.9134233 0.0124   0.0155  4    0.032507547 0.1152797 -0.05026461</span></span>
<span><span class="co">#&gt; Ftl1     85.9937477 0.6847   0.0037  5    0.011745445 0.3276667 -0.30417583</span></span>
<span><span class="co">#&gt; Mif    2439.8454192 0.0036   0.0004  6             NA        NA          NA</span></span>
<span><span class="co">#&gt;        thetaCorrected alphaCorrected betaCorrected alphaSmoothed betaSmoothed</span></span>
<span><span class="co">#&gt; Stmn1     0.007318549   1.347808e+02     1.8582919   131.7760754    1.8168640</span></span>
<span><span class="co">#&gt; Uchl1     0.000000000   1.735938e+05   134.0358378            NA           NA</span></span>
<span><span class="co">#&gt; Hmgb1     0.045258514   2.158710e+01     0.5081917    37.7996053    0.8898576</span></span>
<span><span class="co">#&gt; Gm9794    0.028084312   4.415276e-01    35.1655400     0.3814499   30.3806375</span></span>
<span><span class="co">#&gt; Ftl1      0.009350336   7.322732e+01    33.7207138    58.2949384   26.8444488</span></span>
<span><span class="co">#&gt; Mif       0.000400000   8.830451e+00  2439.8454192            NA           NA</span></span>
<span><span class="co">#&gt;        theta_common ci_upper2   ci_lower2</span></span>
<span><span class="co">#&gt; Stmn1   0.007485426 0.1367155 -0.12174465</span></span>
<span><span class="co">#&gt; Uchl1   0.008250289 0.1712196 -0.15471906</span></span>
<span><span class="co">#&gt; Hmgb1   0.025846831 0.1106315 -0.05893784</span></span>
<span><span class="co">#&gt; Gm9794  0.032507547 0.1152797 -0.05026461</span></span>
<span><span class="co">#&gt; Ftl1    0.011745445 0.3276667 -0.30417583</span></span>
<span><span class="co">#&gt; Mif     0.015817638 0.1062152 -0.07457996</span></span></code></pre></div>
<p>Visualizing model fit. The population of genes with extremely low
dispersion levels forms a separate cluster. We estimate that those genes
have
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>θ</mi><mo>&lt;</mo><mn>0.001</mn></mrow><annotation encoding="application/x-tex">\theta &lt; 0.001</annotation></semantics></math>
and those genes are excluded from the shrinkage procedure by setting
parameter thetaFilter = 0.001 in the correct_theta command.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">celltypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Cortical neurons"</span>, <span class="st">"Gliogenic RGCs"</span>, <span class="st">"IPCs"</span>, <span class="st">"OPCs"</span>, <span class="st">"RGCs"</span><span class="op">)</span></span>
<span><span class="va">p_disp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_disp_fit_theta.html">plot_disp_fit_theta</a></span><span class="op">(</span><span class="va">p</span>, midpoint <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>                                          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">q</span>, <span class="st">" Cast_B6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1e-03</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                 <span class="va">shrunk_estims_vardelta</span>, <span class="va">celltypes</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_disp</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 39 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 71 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 74 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 54 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 84 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-14-1.png" width="960"></p>
<p>Visualizing original and shrunk
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>θ</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p_disp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_disp.html">plot_disp</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">+</span></span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">q</span>, <span class="st">" Cast_B6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1e-03</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                 <span class="va">shrunk_estims_vardelta</span>, <span class="va">celltypes</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_disp</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-15-1.png" width="960">
Alternatively, we can set
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>δ</mi><annotation encoding="application/x-tex">\delta</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>N</mi><annotation encoding="application/x-tex">N</annotation></semantics></math>
parameters manually.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shrunk_estims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_init_params</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/correct_theta.html">correct_theta</a></span><span class="op">(</span><span class="va">q</span>, delta_set <span class="op">=</span> <span class="fl">50</span>, N_set <span class="op">=</span> <span class="fl">30</span>, thetaFilter <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_disp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_disp.html">plot_disp</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">+</span></span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">q</span>, <span class="st">" Cast_B6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1e-03</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                 <span class="va">shrunk_estims</span>, <span class="va">celltypes</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_disp</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-17-1.png" width="960">
### Estimating global beta-binomial parameters</p>
<p>Evaluating global beta-binomial parameters helps to assess a degree
of bias towards the reference allele. This is done by estimating
beta-binomial distribution parameters on all genes, excluding those on
sex chromosomes and the imprinted genes. For the GRCm38 (mm10) mouse
genome, the list of genes on X and Y chromosomes and some validated
imprinted genes are included with the package. These genes are specified
through genes.excl parameter in glob_disp function.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">load_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mm10_genesXY.txt"</span>, package <span class="op">=</span> <span class="st">"ASPEN"</span><span class="op">)</span></span>
<span><span class="va">genesXY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">load_file</span><span class="op">)</span></span>
<span><span class="va">load_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mm10_imprinted_genes.xlsx"</span>, package <span class="op">=</span> <span class="st">"ASPEN"</span><span class="op">)</span></span>
<span><span class="va">genesIMPR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">read.xlsx</a></span><span class="op">(</span><span class="va">load_file</span>, colNames <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">genes2remove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">genesXY</span><span class="op">$</span><span class="va">V1</span>, <span class="va">genesIMPR</span><span class="op">$</span><span class="va">imprinted.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="va">global_estims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span>  <span class="fu"><a href="../reference/glob_disp.html">glob_disp</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, genes.excl <span class="op">=</span> <span class="va">genes2remove</span>, min_counts <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>                        <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">global_estims</span></span>
<span><span class="co">#&gt; $`Cortical neurons`</span></span>
<span><span class="co">#&gt;         mu      theta      alpha       beta </span></span>
<span><span class="co">#&gt;  0.5445001 15.0586614  8.1994433  6.8592218 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Gliogenic RGCs`</span></span>
<span><span class="co">#&gt;         mu      theta      alpha       beta </span></span>
<span><span class="co">#&gt;  0.5307348 19.8779514 10.5499393  9.3280478 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $IPCs</span></span>
<span><span class="co">#&gt;        mu     theta     alpha      beta </span></span>
<span><span class="co">#&gt;  0.541977 18.096688  9.808002  8.288720 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OPCs</span></span>
<span><span class="co">#&gt;        mu     theta     alpha      beta </span></span>
<span><span class="co">#&gt;  0.528725 20.121835 10.638866  9.482886 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $RGCs</span></span>
<span><span class="co">#&gt;         mu      theta      alpha       beta </span></span>
<span><span class="co">#&gt;  0.5298265 19.8136680 10.4979006  9.3159135</span></span></code></pre></div>
<p>Visualizing global allelic ratio distribution across all genes.
Deviation from the balanced allelic expression
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mi>R</mi><mo>=</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">AR = 0.5</annotation></semantics></math>)
indicates a presence of a skew towards the reference allele. If
reference allele bias is identified, the null hypothesis for the allelic
imbalance testing will be adjusted accordingly. For analysis of CastB6
hybrids, instead of using
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><mi>μ</mi><mo>=</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">H_0: \mu = 0.5</annotation></semantics></math>,
the null hypotheses will be adjusted to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><mi>μ</mi><mo>=</mo><mn>0.54</mn></mrow><annotation encoding="application/x-tex">H_0: \mu = 0.54</annotation></semantics></math>
for Cortical neurons and IPCs datasets and to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><mi>μ</mi><mo>=</mo><mn>0.53</mn></mrow><annotation encoding="application/x-tex">H_0: \mu = 0.53</annotation></semantics></math>
for Gliogenic RGCs, OPCs and RGCs.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_glob</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span>,<span class="va">r</span>,<span class="va">s</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_glob_params.html">plot_glob_params</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, <span class="va">r</span>, min_counts <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">s</span>, <span class="st">" Cast_B6"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                 <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, <span class="va">global_estims</span>, <span class="va">celltypes</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: reshape2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'reshape2'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:data.table':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     dcast, melt</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_glob</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-19-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="allelic-imbalance-test-1">Allelic imbalance test<a class="anchor" aria-label="anchor" href="#allelic-imbalance-test-1"></a>
</h4>
<p>Running beta_binom_test function to identify genes with allelic ratio
deviating from the null hypothesis.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_test_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, <span class="va">r</span>, <span class="va">s</span><span class="op">)</span> <span class="fu"><a href="../reference/beta_binom_test.html">beta_binom_test</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, <span class="va">r</span>, min_cells <span class="op">=</span> <span class="fl">5</span>, glob_params <span class="op">=</span> <span class="va">s</span><span class="op">)</span>, </span>
<span>                        <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, <span class="va">shrunk_estims</span>, <span class="va">global_estims</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
beta_binom_test function performs two tests: on the allelic imbalance
and on the allelic variance (deviation from the expected variation for
the genes with similar expression). The output of this function is a
table that combines the output of the estim_bbparams and correct_theta
functions and adds the following columns:\
<p>\</p>
<p>For genes in which the quality cut-off threshold (here we used a
minimum of five cells with at least of five mapped reads) was not met,
the inference is not performed, Those genes have NA values in the fields
specified above. We remove those genes and calculate fdr’s.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_test_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_test_res</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="va">q</span> <span class="op">&lt;-</span> <span class="va">q</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">q</span><span class="op">$</span><span class="va">pval_mean</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#calculating fdr</span></span>
<span><span class="va">bb_test_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_test_res</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="op">{</span><span class="va">q</span><span class="op">$</span><span class="va">fdr_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">q</span><span class="op">$</span><span class="va">pval_mean</span>, method <span class="op">=</span> <span class="st">"fdr"</span><span class="op">)</span>;</span>
<span>                                                <span class="va">q</span><span class="op">$</span><span class="va">fdr_disp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">q</span><span class="op">$</span><span class="va">pval_disp</span>, method <span class="op">=</span> <span class="st">"fdr"</span><span class="op">)</span>;</span>
<span>                                                <span class="va">q</span> <span class="op">&lt;-</span> <span class="va">q</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">q</span><span class="op">$</span><span class="va">fdr_mean</span><span class="op">)</span>,<span class="op">]</span>;</span>
<span>                                                <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Top genes with significant allelic imbalance based on fdr</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_test_res</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="va">q</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AR"</span>, <span class="st">"fdr_mean"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`Cortical neurons`</span></span>
<span><span class="co">#&gt;                 AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1  0.984148582        0</span></span>
<span><span class="co">#&gt; Uchl1  0.998929294        0</span></span>
<span><span class="co">#&gt; Hmgb1  0.976436130        0</span></span>
<span><span class="co">#&gt; Gm9794 0.011239671        0</span></span>
<span><span class="co">#&gt; Ftl1   0.694541696        0</span></span>
<span><span class="co">#&gt; Mif    0.003891258        0</span></span>
<span><span class="co">#&gt; Rps15  0.837685781        0</span></span>
<span><span class="co">#&gt; Rps24  0.203409667        0</span></span>
<span><span class="co">#&gt; Sec61g 0.024583584        0</span></span>
<span><span class="co">#&gt; Eif4a1 0.998208469        0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`Gliogenic RGCs`</span></span>
<span><span class="co">#&gt;                 AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1  0.995680152        0</span></span>
<span><span class="co">#&gt; Uchl1  0.999465812        0</span></span>
<span><span class="co">#&gt; Gm9794 0.008240608        0</span></span>
<span><span class="co">#&gt; Ftl1   0.680179015        0</span></span>
<span><span class="co">#&gt; Rps15  0.864514339        0</span></span>
<span><span class="co">#&gt; Rps24  0.202708166        0</span></span>
<span><span class="co">#&gt; Sec61g 0.032801669        0</span></span>
<span><span class="co">#&gt; Rpl35a 0.999114844        0</span></span>
<span><span class="co">#&gt; Ubb    0.998771109        0</span></span>
<span><span class="co">#&gt; Mrps33 0.027648076        0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $IPCs</span></span>
<span><span class="co">#&gt;                  AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1   0.988831009        0</span></span>
<span><span class="co">#&gt; Uchl1   0.999651163        0</span></span>
<span><span class="co">#&gt; Gm9794  0.011096375        0</span></span>
<span><span class="co">#&gt; Mif     0.002981515        0</span></span>
<span><span class="co">#&gt; Rps15   0.865044891        0</span></span>
<span><span class="co">#&gt; Rps24   0.206429375        0</span></span>
<span><span class="co">#&gt; Sec61g  0.024171380        0</span></span>
<span><span class="co">#&gt; Eif4a1  0.998726346        0</span></span>
<span><span class="co">#&gt; Hnrnpa1 0.026154401        0</span></span>
<span><span class="co">#&gt; Rpl35a  0.996655475        0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OPCs</span></span>
<span><span class="co">#&gt;                 AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1  0.990471159        0</span></span>
<span><span class="co">#&gt; Uchl1  0.999502700        0</span></span>
<span><span class="co">#&gt; Hmgb1  0.947855448        0</span></span>
<span><span class="co">#&gt; Gm9794 0.013659644        0</span></span>
<span><span class="co">#&gt; Ftl1   0.665984639        0</span></span>
<span><span class="co">#&gt; Mif    0.002498002        0</span></span>
<span><span class="co">#&gt; Rps15  0.882669209        0</span></span>
<span><span class="co">#&gt; Rps24  0.207723347        0</span></span>
<span><span class="co">#&gt; Sec61g 0.029949202        0</span></span>
<span><span class="co">#&gt; Eif4a1 0.999097098        0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $RGCs</span></span>
<span><span class="co">#&gt;                 AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1  0.991590207        0</span></span>
<span><span class="co">#&gt; Uchl1  0.998704663        0</span></span>
<span><span class="co">#&gt; Hmgb1  0.950060226        0</span></span>
<span><span class="co">#&gt; Gm9794 0.004747001        0</span></span>
<span><span class="co">#&gt; Ftl1   0.675717150        0</span></span>
<span><span class="co">#&gt; Mif    0.004378284        0</span></span>
<span><span class="co">#&gt; Rps15  0.867951462        0</span></span>
<span><span class="co">#&gt; Rps24  0.196448413        0</span></span>
<span><span class="co">#&gt; Sec61g 0.028287041        0</span></span>
<span><span class="co">#&gt; Eif4a1 0.998527246        0</span></span></code></pre></div>
<p>Visualizing allelic distribution of some of the top genes. We’re
using all cells for the plots below and they’re coloured by log(mean
expression).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#specifiying genes for plotting</span></span>
<span><span class="va">genes_select</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Stmn1"</span>, <span class="st">"Ftl1"</span>, <span class="st">"Mif"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#generating data frame for plotting</span></span>
<span><span class="va">plot_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">genes_select</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/makedf.html">makedf</a></span><span class="op">(</span><span class="va">a1_mat</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">tot_mat</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, gene <span class="op">=</span> <span class="va">q</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_ar_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span>,<span class="va">r</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_distr.html">plot_distr</a></span><span class="op">(</span><span class="va">p</span>, gene <span class="op">=</span> <span class="va">q</span><span class="op">)</span>,</span>
<span>                 <span class="va">plot_data</span>, <span class="va">genes_select</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_ar_dist</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-23-1.png" width="720"></p>
<p>To assess the overall test results, we can plot allelic ratio over
the log (mean expression). This plot has a similar concept to an MA plot
- instead of the log2 fold changes, we plot allelic ratio on the y-axis.
The plot is centered around the global allelic mean, which was used as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>H</mi><mn>0</mn></msub><annotation encoding="application/x-tex">H_0</annotation></semantics></math>
in the test, and as we previously identified, for all cell types in
CastB6 hybrids dataset it deviates from
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mi>R</mi><mo>=</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">AR = 0.5</annotation></semantics></math>.
Genes with significant deviation from the balanced expression (by
default defined as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><msub><mi>g</mi><mn>2</mn></msub><mi>F</mi><mi>C</mi><mo>≥</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">log_2FC \geqslant 1</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>f</mi><mi>d</mi><mi>r</mi><mo>&lt;</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">fdr &lt; 0.05</annotation></semantics></math>)
are indicated in red. Number of such genes biased to either of the
alleles is also shown.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MA_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_MA.html">plot_MA</a></span><span class="op">(</span><span class="va">p</span>, fdr_var <span class="op">=</span> <span class="cn">FALSE</span>,  fdr_cutoff <span class="op">=</span> <span class="fl">0.05</span>, min_logFC <span class="op">=</span> <span class="fl">1</span>, <span class="va">q</span><span class="op">)</span>, </span>
<span>                 <span class="va">bb_test_res</span>, <span class="va">global_estims</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">MA_plot</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-24-1.png" width="960"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Veronika Petrova, Emily Wong.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
