<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>allelic_imbalance • ASPEN</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="allelic_imbalance">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ASPEN</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/allelic_imbalance.html">allelic_imbalance</a></li>
    <li><a class="dropdown-item" href="../articles/allelic_variance.html">allelic_variance</a></li>
    <li><a class="dropdown-item" href="../articles/differential_ASE.html">differential_ASE</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>allelic_imbalance</h1>
            
      

      <div class="d-none name"><code>allelic_imbalance.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="identifying-genes-with-allelic-imbalance">Identifying genes with allelic imbalance<a class="anchor" aria-label="anchor" href="#identifying-genes-with-allelic-imbalance"></a>
</h2>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>ASPEN is a suite of statistical tests for evaluating allele-specific
expression (ASE). It models the allelic ratio — defined as the fraction
of reads from the reference allele relative to the total number of reads
mapped to a gene — using the beta-binomial distribution. The allelic
ratio is restricted to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[0,1]</annotation></semantics></math>
interval and its distribution can be described with beta-binomial mean,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>μ</mi><annotation encoding="application/x-tex">\mu</annotation></semantics></math>,
and dispersion,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>θ</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>.</p>
<p>As with other sequencing-based measurements, single-cell RNA-seq
counts exhibit mean–variance bias: lowly expressed genes tend to display
higher relative variability than highly expressed genes. ASPEN mitigates
this bias by modelling allelic dispersion,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>θ</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>,
as a function of gene expression. Useing the hierarchical Bayes
approach, it shrinks the original dispersion estimates toward the common
trend, which represents the level of dispersion expected for genes with
similar expression.</p>
<p>ASPEN requires two input matrices: the reference allele counts and
the total counts (a sum of counts from both alleles).</p>
The main steps in the workflow are as follows:
</div>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#loading required libraries</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ewonglab.github.io/ASPEN/">ASPEN</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ycphs.github.io/openxlsx/index.html" class="external-link">openxlsx</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://hughjonesd.github.io/huxtable/" class="external-link">huxtable</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="loading-allele-specifc-count-data">Loading allele-specifc count data<a class="anchor" aria-label="anchor" href="#loading-allele-specifc-count-data"></a>
</h3>
<p>In this example, we use mouse brain organoid data from Bl6Cast F1
hybrids (Medina-Cano et al., 2025). We load the reference allele counts
(Bl6 counts) and the total counts (sum of counts from both alleles).
Cell identities were previously annotated based on marker gene
expression, resulting in five major cell types: neurogenic progenitor
cells (adial glial cells, RGCs), intermediate progenitors cells (IPCs),
deep layer neurons (cortical neurons), gliogenic progenitor cells
(gliogenic RGCs) and olygodendrocyte precursor cells (OPCs).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"Bl6_Cast_a1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"Bl6_Cast_tot"</span><span class="op">)</span></span>
<span><span class="va">load_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Bl6_Cast_cell_annot.xlsx"</span>, package <span class="op">=</span> <span class="st">"ASPEN"</span><span class="op">)</span></span>
<span><span class="va">cell_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">read.xlsx</a></span><span class="op">(</span><span class="va">load_file</span>, rowNames <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#adding barcode id</span></span>
<span><span class="va">cell_annot</span><span class="op">$</span><span class="va">cell_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">cell_annot</span><span class="op">$</span><span class="va">clone</span>, <span class="va">cell_annot</span><span class="op">$</span><span class="va">cell_barcode</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/to_md.html" class="external-link">print_md</a></span><span class="op">(</span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/as_huxtable.html" class="external-link">as_huxtable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cell_annot</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; -----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt;  cell_barcode background cell_type    cell_idents  clone  cell_id      </span></span>
<span><span class="co">#&gt; ------------- ---------- ------------ ------------ ------ -------------</span></span>
<span><span class="co">#&gt;  AAACCCACAGCA cast_b6    Gliogenic    Gliogenic    clone3 clone3_AAACC </span></span>
<span><span class="co">#&gt;  GATG                    progenitor   RGCs                CACAGCAGATG  </span></span>
<span><span class="co">#&gt;                          cells                                         </span></span>
<span><span class="co">#&gt;                                                                        </span></span>
<span><span class="co">#&gt;  AAACCCATCAGA cast_b6    Gliogenic    Gliogenic    clone3 clone3_AAACC </span></span>
<span><span class="co">#&gt;  GCGA                    progenitor   RGCs                CATCAGAGCGA  </span></span>
<span><span class="co">#&gt;                          cells                                         </span></span>
<span><span class="co">#&gt;                                                                        </span></span>
<span><span class="co">#&gt;  AAACGAAGTAGT cast_b6    Neurogenic   RGCs         clone3 clone3_AAACG </span></span>
<span><span class="co">#&gt;  TAGA                    progenitor                       AAGTAGTTAGA  </span></span>
<span><span class="co">#&gt;                          cells                                         </span></span>
<span><span class="co">#&gt;                                                                        </span></span>
<span><span class="co">#&gt;  AAACGAAGTGTT cast_b6    Intermediate IPCs         clone3 clone3_AAACG </span></span>
<span><span class="co">#&gt;  ATCG                    neuronal                         AAGTGTTATCG  </span></span>
<span><span class="co">#&gt;                          progenitors                                   </span></span>
<span><span class="co">#&gt;                                                                        </span></span>
<span><span class="co">#&gt;  AAACGAATCTCG cast_b6    Deep layer   Cortical     clone3 clone3_AAACG </span></span>
<span><span class="co">#&gt;  TCAC                    neurons      neurons             AATCTCGTCAC  </span></span>
<span><span class="co">#&gt;                                                                        </span></span>
<span><span class="co">#&gt;  AAAGGATGTATG cast_b6    Intermediate IPCs         clone3 clone3_AAAGG </span></span>
<span><span class="co">#&gt;  CTTG                    neuronal                         ATGTATGCTTG  </span></span>
<span><span class="co">#&gt;                          progenitors                                   </span></span>
<span><span class="co">#&gt; -----------------------------------------------------------------------</span></span></code></pre></div>
<p>The cell abundance for each neuronal subtype</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/to_md.html" class="external-link">print_md</a></span><span class="op">(</span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/as_huxtable.html" class="external-link">as_huxtable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cell_annot</span><span class="op">$</span><span class="va">cell_idents</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; -----------------------------</span></span>
<span><span class="co">#&gt;                   V1         </span></span>
<span><span class="co">#&gt; ----------------- -----------</span></span>
<span><span class="co">#&gt;  Cortical neurons 1615       </span></span>
<span><span class="co">#&gt;                              </span></span>
<span><span class="co">#&gt;    Gliogenic RGCs 1157       </span></span>
<span><span class="co">#&gt;                              </span></span>
<span><span class="co">#&gt;              IPCs 1274       </span></span>
<span><span class="co">#&gt;                              </span></span>
<span><span class="co">#&gt;              OPCs 1523       </span></span>
<span><span class="co">#&gt;                              </span></span>
<span><span class="co">#&gt;              RGCs 935        </span></span>
<span><span class="co">#&gt; -----------------------------</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="counts-normalisation">Counts normalisation<a class="anchor" aria-label="anchor" href="#counts-normalisation"></a>
</h3>
<p>We first normalize the raw single-cell counts using the
[<code><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors()</a></code>] function from <code>scran</code>
package (Lun, et al. 2016). We then create a
<code>SingleCellExperiment</code> object using the total and reference
allele count matrices.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#keeping annotated cells in the count matrices</span></span>
<span><span class="va">Cast_B6_a1</span> <span class="op">&lt;-</span> <span class="va">Cast_B6_a1</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*_"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Cast_B6_a1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cell_annot</span><span class="op">$</span><span class="va">cell_barcode</span><span class="op">]</span></span>
<span><span class="va">Cast_B6_tot</span> <span class="op">&lt;-</span> <span class="va">Cast_B6_tot</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".*_"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">Cast_B6_tot</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cell_annot</span><span class="op">$</span><span class="va">cell_barcode</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#creating SingleCellExperiment object</span></span>
<span><span class="va">ase_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Cast_B6_a1</span><span class="op">)</span>,</span>
<span>                                              tot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Cast_B6_tot</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Lowly expressed genes (expressed in less than 10 cells) are
removed.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#removing lowly expressed genes</span></span>
<span><span class="va">ase_sce</span> <span class="op">&lt;-</span> <span class="va">ase_sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">ase_sce</span><span class="op">)</span><span class="op">[[</span><span class="st">'tot'</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/to_md.html" class="external-link">print_md</a></span><span class="op">(</span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/as_huxtable.html" class="external-link">as_huxtable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ase_sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; ----------------------</span></span>
<span><span class="co">#&gt;                    V1 </span></span>
<span><span class="co">#&gt;                       </span></span>
<span><span class="co">#&gt;                  1551 </span></span>
<span><span class="co">#&gt;                       </span></span>
<span><span class="co">#&gt;                  6312 </span></span>
<span><span class="co">#&gt; ----------------------</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#adding sample id to the metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">ase_sce</span><span class="op">)</span><span class="op">$</span><span class="va">replicate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_.*"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">ase_sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#calculate size factors</span></span>
<span><span class="va">ase_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">ase_sce</span>, </span>
<span>                             clusters<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">ase_sce</span><span class="op">)</span><span class="op">$</span><span class="va">replicate</span>, assay.type <span class="op">=</span> <span class="st">"tot"</span><span class="op">)</span></span></code></pre></div>
<p>The reference allele and total counts are normalized in parallel
using the same size factor estimates.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#normalizing counts</span></span>
<span><span class="va">ase_sce</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">ase_sce</span>, </span>
<span>                          size.factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">ase_sce</span><span class="op">)</span><span class="op">$</span><span class="va">sizeFactor</span>,</span>
<span>                          log <span class="op">=</span> <span class="cn">NULL</span>, transform <span class="op">=</span> <span class="st">"none"</span>, assay.type <span class="op">=</span> <span class="st">"tot"</span>, name <span class="op">=</span> <span class="st">"tot_norm"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#normalizing reference counts by the same size factors</span></span>
<span><span class="va">ase_sce</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">ase_sce</span>, </span>
<span>                          size.factors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">ase_sce</span><span class="op">)</span><span class="op">$</span><span class="va">sizeFactor</span>,</span>
<span>                          log <span class="op">=</span> <span class="cn">NULL</span>, transform <span class="op">=</span> <span class="st">"none"</span>, assay.type <span class="op">=</span> <span class="st">"a1"</span>, name <span class="op">=</span> <span class="st">"a1_norm"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#checking that normalised counts assays are added to the SingleCellExperiment object</span></span>
<span><span class="va">ase_sce</span><span class="op">@</span><span class="va">assays</span></span>
<span><span class="co">#&gt; An object of class "SimpleAssays"</span></span>
<span><span class="co">#&gt; Slot "data":</span></span>
<span><span class="co">#&gt; List of length 4</span></span>
<span><span class="co">#&gt; names(4): a1 tot tot_norm a1_norm</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="splitting-count-matrices-by-cell-type">Splitting count matrices by cell type<a class="anchor" aria-label="anchor" href="#splitting-count-matrices-by-cell-type"></a>
</h3>
<p>Allelic imbalance test is run separately for each cell type. First,
we split the metadata object by cell types.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#splitting the metadata by cell type</span></span>
<span><span class="va">cell_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">cell_annot</span>, f <span class="op">=</span> <span class="va">cell_annot</span><span class="op">$</span><span class="va">cell_idents</span><span class="op">)</span></span></code></pre></div>
<p>Since beta-binomial parameters are estimated from raw counts, we
first split the raw total and reference allele count matrices by cell
type.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#splitting SingleCellExperiemnt object by cell types</span></span>
<span><span class="va">ase_sce_byct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cell_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">ase_sce_byct</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">ase_sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ase_sce</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">cell_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cell_id</span><span class="op">]</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">#removing genes with low expression</span></span>
<span><span class="va">ase_sce_byct_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ase_sce_byct</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> </span>
<span>                                          <span class="va">q</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span><span class="op">[[</span><span class="st">'tot'</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span>, <span class="op">]</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">#extracting raw total counts</span></span>
<span><span class="va">tot_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ase_sce_byct_filt</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span><span class="op">[[</span><span class="st">'tot'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#extracting raw reference allele counts</span></span>
<span><span class="va">a1_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ase_sce_byct_filt</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span><span class="op">[[</span><span class="st">'a1'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#selecting genes that matched filtering criteria</span></span>
<span><span class="va">a1_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span><span class="op">)</span> <span class="va">p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span>, <span class="op">]</span>, <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>Before proceeding, we check that the reference and total count
matrices have the same dimensions. This is for demonstration purposes
only. If this condition is not met, ASPEN will issue a warning.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span>, <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt; [1] TRUE TRUE</span></span></code></pre></div>
<p>Next, we check that the gene order is identical between the reference
and total count matrices. Again, this is for demonstration purposes
only. If this condition is not met, ASPEN will issue a warning.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span><span class="op">)</span>, <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1386 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1414 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1503 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1521 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; 1520</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="estimating-beta-binomial-parameters">Estimating beta-binomial parameters<a class="anchor" aria-label="anchor" href="#estimating-beta-binomial-parameters"></a>
</h3>
<p>We estimate beta-binomial parameters separately for each cell
type.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_init_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/estim_bbparams.html">estim_bbparams</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, min_cells <span class="op">=</span> <span class="fl">5</span>, cores <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>, <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
The output is a table with the following columns:
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/to_md.html" class="external-link">print_md</a></span><span class="op">(</span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/as_huxtable.html" class="external-link">as_huxtable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bb_init_params</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in to_md.huxtable(ht, ...): Couldn't print whole table in max_width = 80 characters.</span></span>
<span><span class="co">#&gt; Printing 8/9 columns.</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt;         N     AR tot_gene tot_gene    alpha     beta  bb_mu bb_theta </span></span>
<span><span class="co">#&gt;                     _mean _varianc                                   </span></span>
<span><span class="co">#&gt;                                  e                                   </span></span>
<span><span class="co">#&gt; --------- ------ -------- -------- -------- -------- ------ -------- </span></span>
<span><span class="co">#&gt;       850  0.984     5.94     29.3      175     2.41  0.986   0.0056 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;  1.06e+03  0.999     8.21     45.4 4.41e+03     3.42  0.999   0.0002 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;        37  0.976     1.09     1.78     14.6    0.343  0.977    0.067 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;        25 0.0113    0.953     1.54    0.807     64.2 0.0124   0.0154 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;  1.43e+03  0.695     21.3      322      184     84.5  0.685   0.0037 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;        93 0.0039      1.4     3.57     25.2 6.94e+03 0.0036   0.0001 </span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="defining-lowly-expressed-genes">Defining lowly expressed genes<a class="anchor" aria-label="anchor" href="#defining-lowly-expressed-genes"></a>
</h3>
<p>ASPEN applies shrinkage selectively - genes with very low dispersion
are not moderated and their allelic imbalance is evaluated using the
unadjusted values. Genes with stable dispersion are determined based on
the residuals from the dispersion modeling step. ASPEN calculates the
meadian absolute deviation-squared
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mrow><mi>M</mi><mi>A</mi><mi>D</mi></mrow><mn>2</mn></msup><annotation encoding="application/x-tex">{MAD}^2</annotation></semantics></math>),
which is used as a cut-off.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">min_cutoff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_init_params</span>, <span class="va">calc_mad</span><span class="op">)</span></span>
<span><span class="va">min_cutoff</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 0.002254469</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] 0.001084665</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [1] 0.002810162</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; [1] 0.00113423</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt; [1] 0.001276519</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="estimate-appropriate-shrinkage-parameters">Estimate appropriate shrinkage parameters<a class="anchor" aria-label="anchor" href="#estimate-appropriate-shrinkage-parameters"></a>
</h3>
<p>We estimate shrinkage parameters,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>N</mi><annotation encoding="application/x-tex">N</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>δ</mi><annotation encoding="application/x-tex">\delta</annotation></semantics></math>,
separately for each cell type. Optionally, genes with very low
dispersion can be excluded from the estimation by setting a minimum
cut-off value with <code>thetaFilter</code> parameter</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1001011</span><span class="op">)</span></span>
<span><span class="va">shrink_pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/estim_delta.html">estim_delta</a></span><span class="op">(</span><span class="va">p</span>, thetaFilter <span class="op">=</span> <span class="va">q</span><span class="op">)</span>,</span>
<span>                      <span class="va">bb_init_params</span>, <span class="va">min_cutoff</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">shrink_pars</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    17    18 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    13    15 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    17    17 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    14    17 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt;     N delta </span></span>
<span><span class="co">#&gt;    20    19</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="performing-bayesian-shrinkage">Performing Bayesian shrinkage<a class="anchor" aria-label="anchor" href="#performing-bayesian-shrinkage"></a>
</h3>
<p>We use [<code><a href="../reference/correct_theta.html">correct_theta()</a></code>] from <code>ASPEN</code> to
obtain posterior dispersion estimates. The minimum dispersion value is
set with <code>thetaFilter</code> parameter. and genes with residuals
below that value are excluded shrinkage.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_init_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_init_params</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="va">q</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">q</span><span class="op">$</span><span class="va">bb_theta</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">shrunk_estims_vardelta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/correct_theta.html">correct_theta</a></span><span class="op">(</span><span class="va">p</span>, N_set <span class="op">=</span> <span class="va">q</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, delta_set <span class="op">=</span> <span class="va">q</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, thetaFilter <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span>,</span>
<span>                                 <span class="va">bb_init_params</span>, <span class="va">shrink_pars</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/to_md.html" class="external-link">print_md</a></span><span class="op">(</span><span class="fu"><a href="https://hughjonesd.github.io/huxtable/reference/as_huxtable.html" class="external-link">as_huxtable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">shrunk_estims_vardelta</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in to_md.huxtable(ht, ...): Couldn't print whole table in max_width = 80 characters.</span></span>
<span><span class="co">#&gt; Printing 8/13 columns.</span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span>
<span><span class="co">#&gt;         N     AR tot_gene tot_gene    alpha     beta  bb_mu bb_theta </span></span>
<span><span class="co">#&gt;                     _mean _varianc                                   </span></span>
<span><span class="co">#&gt;                                  e                                   </span></span>
<span><span class="co">#&gt; --------- ------ -------- -------- -------- -------- ------ -------- </span></span>
<span><span class="co">#&gt;       850  0.984     5.94     29.3      175     2.41  0.986   0.0056 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;  1.06e+03  0.999     8.21     45.4 4.41e+03     3.42  0.999   0.0002 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;        37  0.976     1.09     1.78     14.6    0.343  0.977    0.067 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;        25 0.0113    0.953     1.54    0.807     64.2 0.0124   0.0154 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;  1.43e+03  0.695     21.3      322      184     84.5  0.685   0.0037 </span></span>
<span><span class="co">#&gt;                                                                      </span></span>
<span><span class="co">#&gt;        93 0.0039      1.4     3.57     25.2 6.94e+03 0.0036   0.0001 </span></span>
<span><span class="co">#&gt; ---------------------------------------------------------------------</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-model-fit-">Visualizing model fit.<a class="anchor" aria-label="anchor" href="#visualizing-model-fit-"></a>
</h3>
<p>Genes with extremely low dispersion levels form a distinct cluster in
the dispersion–expression plot. We estimate that those genes have
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>θ</mi><mo>&lt;</mo><mn>0.001</mn></mrow><annotation encoding="application/x-tex">\theta &lt; 0.001</annotation></semantics></math>.
To exclude them from the shrinkage procedure, we set
<code>thetaFilter = 0.001</code> in the [<code><a href="../reference/correct_theta.html">correct_theta()</a></code>]
function.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">celltypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Cortical neurons"</span>, <span class="st">"Gliogenic RGCs"</span>, <span class="st">"IPCs"</span>, <span class="st">"OPCs"</span>, <span class="st">"RGCs"</span><span class="op">)</span></span>
<span><span class="va">p_disp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_disp_fit_theta.html">plot_disp_fit_theta</a></span><span class="op">(</span><span class="va">p</span>, midpoint <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">+</span></span>
<span>                                          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">q</span>, <span class="st">" Cast_B6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1e-03</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                 <span class="va">shrunk_estims_vardelta</span>, <span class="va">celltypes</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_disp</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 48 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 73 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 81 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 60 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 96 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_pointdensity()`).</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-17-1.png" width="960"></p>
<p>Visualizing original and shrunk
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>θ</mi><annotation encoding="application/x-tex">\theta</annotation></semantics></math>’s.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p_disp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_disp.html">plot_disp</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">+</span></span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">q</span>, <span class="st">" Cast_B6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1e-03</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                 <span class="va">shrunk_estims_vardelta</span>, <span class="va">celltypes</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_disp</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span>
<span><span class="co">#&gt; Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-18-1.png" width="960"></p>
<p>Alternatively, we can set
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>δ</mi><annotation encoding="application/x-tex">\delta</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>N</mi><annotation encoding="application/x-tex">N</annotation></semantics></math>
parameters manually.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shrunk_estims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_init_params</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/correct_theta.html">correct_theta</a></span><span class="op">(</span><span class="va">q</span>, delta_set <span class="op">=</span> <span class="fl">50</span>, N_set <span class="op">=</span> <span class="fl">30</span>, thetaFilter <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_disp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_disp.html">plot_disp</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">+</span></span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">q</span>, <span class="st">" Cast_B6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1e-03</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, linewidth <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                 <span class="va">shrunk_estims</span>, <span class="va">celltypes</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_disp</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span>
<span><span class="co">#&gt; Removed 1 row containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-20-1.png" width="960">
### Estimating global beta-binomial parameters</p>
<p>Evaluating global beta-binomial parameters provides an overall
measure of bias toward the reference allele. This is done by estimating
beta-binomial distribution parameters across all genes, excluding those
located on sex chromosomes and known imprinted genes. For the GRCm38
(mm10) mouse genome, the package includes predefined lists of X- and
Y-linked genes as well as a set of validated imprinted genes. These can
be excluded by passing them to the <code>genes.excl</code> parameter in
the [<code><a href="../reference/glob_disp.html">glob_disp()</a></code>] function.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">load_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mm10_genesXY.txt"</span>, package <span class="op">=</span> <span class="st">"ASPEN"</span><span class="op">)</span></span>
<span><span class="va">genesXY</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">load_file</span><span class="op">)</span></span>
<span><span class="va">load_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mm10_imprinted_genes.xlsx"</span>, package <span class="op">=</span> <span class="st">"ASPEN"</span><span class="op">)</span></span>
<span><span class="va">genesIMPR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">read.xlsx</a></span><span class="op">(</span><span class="va">load_file</span>, colNames <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">genes2remove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">genesXY</span><span class="op">$</span><span class="va">V1</span>, <span class="va">genesIMPR</span><span class="op">$</span><span class="va">imprinted.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="va">global_estims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span><span class="op">)</span>  <span class="fu"><a href="../reference/glob_disp.html">glob_disp</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, genes.excl <span class="op">=</span> <span class="va">genes2remove</span>, min_counts <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>                        <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">global_estims</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;     mu.1  theta.1  alpha.1   beta.1 </span></span>
<span><span class="co">#&gt; 0.544500 0.066400 8.203912 6.862176 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt;      mu.1   theta.1   alpha.1    beta.1 </span></span>
<span><span class="co">#&gt;  0.530700  0.050300 10.553884  9.332994 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt;     mu.1  theta.1  alpha.1   beta.1 </span></span>
<span><span class="co">#&gt; 0.541900 0.055100 9.827842 8.307707 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt;      mu.1   theta.1   alpha.1    beta.1 </span></span>
<span><span class="co">#&gt;  0.528700  0.049700 10.639340  9.483841 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt;      mu.1   theta.1   alpha.1    beta.1 </span></span>
<span><span class="co">#&gt;  0.529800  0.050500 10.496454  9.314595</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-global-allelic-ratio-distribution-across-all-genes-">Visualizing global allelic ratio distribution across all genes.<a class="anchor" aria-label="anchor" href="#visualizing-global-allelic-ratio-distribution-across-all-genes-"></a>
</h3>
<p>Deviation from the balanced allelic expression
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>A</mi><mi>R</mi><mo>=</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">AR = 0.5</annotation></semantics></math>)
indicates a presence of a skew towards the reference allele. If
reference allele bias is detected, the null hypothesis for the allelic
imbalance testing should be adjusted accordingly. For analysis of the
Bl6Cast hybrids, instead of using
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><mi>μ</mi><mo>=</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">H_0: \mu = 0.5</annotation></semantics></math>,
the null hypotheses will be adjusted to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><mi>μ</mi><mo>=</mo><mn>0.54</mn></mrow><annotation encoding="application/x-tex">H_0: \mu = 0.54</annotation></semantics></math>
for Cortical neurons and IPCs datasets and to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><mi>μ</mi><mo>=</mo><mn>0.53</mn></mrow><annotation encoding="application/x-tex">H_0: \mu = 0.53</annotation></semantics></math>
for Gliogenic RGCs, OPCs and RGCs.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_glob</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span>,<span class="va">r</span>,<span class="va">s</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_glob_params.html">plot_glob_params</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, <span class="va">r</span>, min_counts <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>                               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">s</span>, <span class="st">" Cast_B6"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                 <span class="va">a1_mat</span>, <span class="va">tot_mat</span>, <span class="va">global_estims</span>, <span class="va">celltypes</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The melt generic in data.table has been passed a data.frame and will</span></span>
<span><span class="co">#&gt; attempt to redirect to the relevant reshape2 method; please note that reshape2</span></span>
<span><span class="co">#&gt; is superseded and is no longer actively developed, and this redirection is now</span></span>
<span><span class="co">#&gt; deprecated. To continue using melt methods from reshape2 while both libraries</span></span>
<span><span class="co">#&gt; are attached, e.g. melt.list, you can prepend the namespace, i.e.</span></span>
<span><span class="co">#&gt; reshape2::melt(plot_data). In the next version, this warning will become an</span></span>
<span><span class="co">#&gt; error.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: The melt generic in data.table has been passed a data.frame and will</span></span>
<span><span class="co">#&gt; attempt to redirect to the relevant reshape2 method; please note that reshape2</span></span>
<span><span class="co">#&gt; is superseded and is no longer actively developed, and this redirection is now</span></span>
<span><span class="co">#&gt; deprecated. To continue using melt methods from reshape2 while both libraries</span></span>
<span><span class="co">#&gt; are attached, e.g. melt.list, you can prepend the namespace, i.e.</span></span>
<span><span class="co">#&gt; reshape2::melt(plot_data). In the next version, this warning will become an</span></span>
<span><span class="co">#&gt; error.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: The melt generic in data.table has been passed a data.frame and will</span></span>
<span><span class="co">#&gt; attempt to redirect to the relevant reshape2 method; please note that reshape2</span></span>
<span><span class="co">#&gt; is superseded and is no longer actively developed, and this redirection is now</span></span>
<span><span class="co">#&gt; deprecated. To continue using melt methods from reshape2 while both libraries</span></span>
<span><span class="co">#&gt; are attached, e.g. melt.list, you can prepend the namespace, i.e.</span></span>
<span><span class="co">#&gt; reshape2::melt(plot_data). In the next version, this warning will become an</span></span>
<span><span class="co">#&gt; error.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: The melt generic in data.table has been passed a data.frame and will</span></span>
<span><span class="co">#&gt; attempt to redirect to the relevant reshape2 method; please note that reshape2</span></span>
<span><span class="co">#&gt; is superseded and is no longer actively developed, and this redirection is now</span></span>
<span><span class="co">#&gt; deprecated. To continue using melt methods from reshape2 while both libraries</span></span>
<span><span class="co">#&gt; are attached, e.g. melt.list, you can prepend the namespace, i.e.</span></span>
<span><span class="co">#&gt; reshape2::melt(plot_data). In the next version, this warning will become an</span></span>
<span><span class="co">#&gt; error.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: The melt generic in data.table has been passed a data.frame and will</span></span>
<span><span class="co">#&gt; attempt to redirect to the relevant reshape2 method; please note that reshape2</span></span>
<span><span class="co">#&gt; is superseded and is no longer actively developed, and this redirection is now</span></span>
<span><span class="co">#&gt; deprecated. To continue using melt methods from reshape2 while both libraries</span></span>
<span><span class="co">#&gt; are attached, e.g. melt.list, you can prepend the namespace, i.e.</span></span>
<span><span class="co">#&gt; reshape2::melt(plot_data). In the next version, this warning will become an</span></span>
<span><span class="co">#&gt; error.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_glob</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-22-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="allelic-imbalance-test">Allelic imbalance test<a class="anchor" aria-label="anchor" href="#allelic-imbalance-test"></a>
</h3>
<p>The allelic imbalance test is performed on the normalized counts.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#extracting normalised reference allele and total counts</span></span>
<span><span class="va">tot_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ase_sce_byct_filt</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span><span class="op">[[</span><span class="st">'tot_norm'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#extracting raw reference allele counts</span></span>
<span><span class="va">a1_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ase_sce_byct_filt</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span><span class="op">[[</span><span class="st">'a1_norm'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#ensure that the genes order is the same</span></span>
<span><span class="va">a1_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span><span class="op">)</span> <span class="va">p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span>, <span class="op">]</span>, <span class="va">a1_norm</span>, <span class="va">tot_norm</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>We run [<code><a href="../reference/bb_mean.html">bb_mean()</a></code>] to identify genes with allelic ratio
deviating from the null hypothesis. By default, ASPEN requires each gene
to be expressed in at least 5 cells with a minimum of 5 counts per cell.
These thresholds can be adjusted with <code>min_cells</code> and
<code>min_counts</code> parameters.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_mean_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, <span class="va">r</span>, <span class="va">s</span><span class="op">)</span> <span class="fu"><a href="../reference/bb_mean.html">bb_mean</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">q</span>, <span class="va">r</span>, </span>
<span>                                                   min_cells <span class="op">=</span> <span class="fl">5</span>, min_counts <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                                   glob_params <span class="op">=</span> <span class="va">s</span><span class="op">)</span>, </span>
<span>                      <span class="va">a1_norm</span>, <span class="va">tot_norm</span>, <span class="va">shrunk_estims</span>, <span class="va">global_estims</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
The output of [<code>beta_binom_test()</code>] function is a table that
combines the output of the [<code><a href="../reference/estim_bbparams.html">estim_bbparams()</a></code>] and
[<code><a href="../reference/correct_theta.html">correct_theta()</a></code>] functions and adds the following
columns:\
<p>\</p>
<p>If a gene does not meet the quality filter (here, a minimum of 5
cells with at least 5 mapped reads), the allelic imbalance test is not
performed for that gene. These genes will have <code>NA</code> in the
above fields. We remove such genes before calculating the false
discovery rate (FDR).</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_mean_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_mean_res</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="va">q</span> <span class="op">&lt;-</span> <span class="va">q</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">q</span><span class="op">$</span><span class="va">pval_mean</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#calculating fdr</span></span>
<span><span class="va">bb_mean_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_mean_res</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="op">{</span><span class="va">q</span><span class="op">$</span><span class="va">fdr_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">q</span><span class="op">$</span><span class="va">pval_mean</span>, </span>
<span>                                                                       method <span class="op">=</span> <span class="st">"fdr"</span><span class="op">)</span>;                                                          <span class="va">q</span> <span class="op">&lt;-</span> <span class="va">q</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">q</span><span class="op">$</span><span class="va">fdr_mean</span><span class="op">)</span>,<span class="op">]</span>;</span>
<span>                                                <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">q</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Top 10 genes with significant allelic imbalance after
FDR-adjustment</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bb_mean_res</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="va">q</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AR"</span>, <span class="st">"fdr_mean"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;                 AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1  0.984126550        0</span></span>
<span><span class="co">#&gt; Uchl1  0.998927886        0</span></span>
<span><span class="co">#&gt; Hmgb1  0.976436130        0</span></span>
<span><span class="co">#&gt; Gm9794 0.011265810        0</span></span>
<span><span class="co">#&gt; Ftl1   0.694522296        0</span></span>
<span><span class="co">#&gt; Mif    0.003895411        0</span></span>
<span><span class="co">#&gt; Rps15  0.838435691        0</span></span>
<span><span class="co">#&gt; Rps24  0.203198897        0</span></span>
<span><span class="co">#&gt; Sec61g 0.024611841        0</span></span>
<span><span class="co">#&gt; Eif4a1 0.998207009        0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt;                  AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1   0.995669976        0</span></span>
<span><span class="co">#&gt; Uchl1   0.999464668        0</span></span>
<span><span class="co">#&gt; Gm9794  0.008240608        0</span></span>
<span><span class="co">#&gt; Ftl1    0.680002167        0</span></span>
<span><span class="co">#&gt; Rps15   0.864274116        0</span></span>
<span><span class="co">#&gt; Rps24   0.202660961        0</span></span>
<span><span class="co">#&gt; Sec61g  0.032801669        0</span></span>
<span><span class="co">#&gt; Hnrnpa1 0.018688305        0</span></span>
<span><span class="co">#&gt; Rpl35a  0.999114066        0</span></span>
<span><span class="co">#&gt; Ubb     0.998768725        0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt;                  AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1   0.988800741        0</span></span>
<span><span class="co">#&gt; Uchl1   0.999650513        0</span></span>
<span><span class="co">#&gt; Gm9794  0.011140701        0</span></span>
<span><span class="co">#&gt; Ftl1    0.672111891        0</span></span>
<span><span class="co">#&gt; Mif     0.002981515        0</span></span>
<span><span class="co">#&gt; Rps15   0.865347111        0</span></span>
<span><span class="co">#&gt; Rps24   0.206824332        0</span></span>
<span><span class="co">#&gt; Sec61g  0.024171380        0</span></span>
<span><span class="co">#&gt; Eif4a1  0.998726346        0</span></span>
<span><span class="co">#&gt; Hnrnpa1 0.026185317        0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt;                 AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1  0.990463015        0</span></span>
<span><span class="co">#&gt; Uchl1  0.999502275        0</span></span>
<span><span class="co">#&gt; Hmgb1  0.947855448        0</span></span>
<span><span class="co">#&gt; Gm9794 0.013659644        0</span></span>
<span><span class="co">#&gt; Ftl1   0.666048789        0</span></span>
<span><span class="co">#&gt; Mif    0.002498002        0</span></span>
<span><span class="co">#&gt; Rps15  0.882687034        0</span></span>
<span><span class="co">#&gt; Rps24  0.207635447        0</span></span>
<span><span class="co">#&gt; Sec61g 0.029949202        0</span></span>
<span><span class="co">#&gt; Eif4a1 0.999096341        0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt;                 AR fdr_mean</span></span>
<span><span class="co">#&gt; Stmn1  0.991578964        0</span></span>
<span><span class="co">#&gt; Uchl1  0.998702983        0</span></span>
<span><span class="co">#&gt; Gm9794 0.004754441        0</span></span>
<span><span class="co">#&gt; Ftl1   0.675546570        0</span></span>
<span><span class="co">#&gt; Mif    0.004378284        0</span></span>
<span><span class="co">#&gt; Rps15  0.867803259        0</span></span>
<span><span class="co">#&gt; Rps24  0.195924785        0</span></span>
<span><span class="co">#&gt; Sec61g 0.028287041        0</span></span>
<span><span class="co">#&gt; Eif4a1 0.998525074        0</span></span>
<span><span class="co">#&gt; Rpl35a 0.997623061        0</span></span></code></pre></div>
<p>We can visualize the allelic distribution for some of the top-ranked
genes. In the plots below, each point represents a single cell, and
points are colored by the log(mean expression).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#specifiying genes for plotting</span></span>
<span><span class="va">genes_select</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Stmn1"</span>, <span class="st">"Ftl1"</span>, <span class="st">"Mif"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#generating data frame for plotting</span></span>
<span><span class="va">plot_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">genes_select</span>, <span class="kw">function</span><span class="op">(</span><span class="va">q</span><span class="op">)</span> <span class="fu"><a href="../reference/makedf.html">makedf</a></span><span class="op">(</span><span class="va">a1_mat</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">tot_mat</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, gene <span class="op">=</span> <span class="va">q</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_ar_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">p</span>,<span class="va">q</span>,<span class="va">r</span><span class="op">)</span> <span class="fu"><a href="../reference/plot_distr.html">plot_distr</a></span><span class="op">(</span><span class="va">p</span>, gene <span class="op">=</span> <span class="va">q</span><span class="op">)</span>,</span>
<span>                 <span class="va">plot_data</span>, <span class="va">genes_select</span>, SIMPLIFY <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">grid.arrange</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">p_ar_dist</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="allelic_imbalance_files/figure-html/unnamed-chunk-27-1.png" width="720"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Veronika Petrova, Muqing Niu, Thomas Vierbuchen, Emily Wong.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
