<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ ASPEN</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ASPEN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/allelic_imbalance.html">allelic_imbalance</a>
    </li>
    <li>
      <a href="../articles/allelic_variance.html">allelic_variance</a>
    </li>
    <li>
      <a href="../articles/dynamic_allelic_changes.html">dynamic_allelic_changes</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="allelicSwitch.html">allelicSwitch()</a></code> </p>
        </td>
        <td><p>Performs a number of test to evaluate changes in allelic distribution between the groups.</p></td>
      </tr><tr><td>
          <p><code><a href="beta_binom_test.html">beta_binom_test()</a></code> </p>
        </td>
        <td><p>Performs likelihood ratio test between H_0: no allelic imbalance (mean allelic ratio = 0.5, or another theoretical value based on the global parameter estimates) and H_1: allelic imbalance is present. Likelihoods for H_0 and H_1 are calculated based on the mu and theta parameters from correct_theta function</p></td>
      </tr><tr><td>
          <p><code><a href="correct_theta.html">correct_theta()</a></code> </p>
        </td>
        <td><p>Fits a local regression model with dispersion as a function of total gene counts. Predicted values are the expected dispersion for genes with similar expression levels The initial dispersion estimates are shrunk towards the common dispersion</p></td>
      </tr><tr><td>
          <p><code><a href="correct_theta_bygroup.html">correct_theta_bygroup()</a></code> </p>
        </td>
        <td><p>Fits a local regression model with dispersion as a function of total gene counts. Predicted values are the expected dispersion for genes with similar expression levels The initial dispersion estimates are shrunk towards the common dispersion</p></td>
      </tr><tr><td>
          <p><code><a href="estim_bbparams.html">estim_bbparams()</a></code> </p>
        </td>
        <td><p>Estimate beta-binomial distribution parameters. This is the initial function that estimates beta-binomial distribution parameters for the allele-specific counts for each gene. The users may filter out low abundant or low expressed genes through quality control parameters of minimum number of reads per cell and minimum number of cells per gene.</p></td>
      </tr><tr><td>
          <p><code><a href="estim_bbparams_bygroup.html">estim_bbparams_bygroup()</a></code> </p>
        </td>
        <td><p>Estimate beta-binomial distribution parameters. This is the initial function that estimates beta-binomial distribution parameters for the allele-specific counts for each gene. The users may filter out low abundant or low expressed genes through quality control parameters of minimum number of reads per cell and minimum number of cells per gene.</p></td>
      </tr><tr><td>
          <p><code><a href="estim_delta.html">estim_delta()</a></code> </p>
        </td>
        <td><p>Estimating appropriate tuning paramater delta and the number of degrees of freedom, N. We assume that dispersion follows Gamma distribution. Appropriate shrinkage is selected based on the MLE of the the difference between fitted dispersion and shrunk dispersion</p></td>
      </tr><tr><td>
          <p><code><a href="geBoxplot.html">geBoxplot()</a></code> </p>
        </td>
        <td><p>Boxplot of allele-specific expression</p></td>
      </tr><tr><td>
          <p><code><a href="glob_disp.html">glob_disp()</a></code> </p>
        </td>
        <td><p>Estimates global beta-binomial distribution parameters on all genes together to evaluate the degree of skewness towards the reference allele. For the estimation, genes located on the sex chromosomes and imprinted genes are excluded</p></td>
      </tr><tr><td>
          <p><code><a href="makedf.html">makedf()</a></code> </p>
        </td>
        <td><p>preparing data frame for plotting</p></td>
      </tr><tr><td>
          <p><code><a href="make_plotdf_simul.html">make_plotdf_simul()</a></code> </p>
        </td>
        <td><p>posterior counts simulated under the expected dispersion</p></td>
      </tr><tr><td>
          <p><code><a href="plot_disp.html">plot_disp()</a></code> </p>
        </td>
        <td><p>Plot to visualize original and shrunk dispersion</p></td>
      </tr><tr><td>
          <p><code><a href="plot_disp_fit_mean.html">plot_disp_fit_mean()</a></code> </p>
        </td>
        <td><p>Plot to visualize allelic ratio modelled as a function of total gene counts</p></td>
      </tr><tr><td>
          <p><code><a href="plot_disp_fit_theta.html">plot_disp_fit_theta()</a></code> </p>
        </td>
        <td><p>Plot to visualize dispersion modelled as a function of total gene counts</p></td>
      </tr><tr><td>
          <p><code><a href="plot_distr.html">plot_distr()</a></code> </p>
        </td>
        <td><p>plotting allelic ratio distribution</p></td>
      </tr><tr><td>
          <p><code><a href="plot_exp_disp.html">plot_exp_disp()</a></code> </p>
        </td>
        <td><p>plotting mean gene expression vs allelic dispersion</p></td>
      </tr><tr><td>
          <p><code><a href="plot_glob_params.html">plot_glob_params()</a></code> </p>
        </td>
        <td><p>Plot histogram of allelic ratio distribution across all genes with beta function based on alpha and beta global parameters</p></td>
      </tr><tr><td>
          <p><code><a href="plot_MA.html">plot_MA()</a></code> </p>
        </td>
        <td><p>Plotting distribution of allelic ratio over mean expression</p></td>
      </tr><tr><td>
          <p><code><a href="plot_theta_density.html">plot_theta_density()</a></code> </p>
        </td>
        <td><p>Density plot comparing allelic ratio distribution under observed and expected dispersion</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Veronika Petrova, Emily Wong.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

